#1211 made sure duplicate features are not used

abhaykatheria opened the pull request:
I made sure even if there are duplicate feature extractors only unique ones are used.

marco-c commented:
> I made sure even if there are duplicate feature extractors only unique ones are used.

It'd be better to have an assertion that fails when there are duplicate features, instead of silently ignoring the duplication.

abhaykatheria commented:
I made a list of all the feature classes that are being used.

`        self.feature_extractors = []
        used_features=[]
        for feature_extractor in feature_extractors:
            for feature in FEATURE_CLASSES:
                if isinstance(feature_extractor,feature):
                    if feature not in used_features:
                        used_features.append(feature)
                        self.feature_extractors.append(feature_extractor)
                    else :
                        assert False,("Duplicate extractors used")`

abhaykatheria commented:
`        self.feature_extractors = []
        used_features=[]
        for feature_extractor in feature_extractors:
            for x in FEATURE_CLASSES:
                if isinstance(feature_extractor,x):
                    if x not in used_features:
                        used_features.append(x)
                        self.feature_extractors.append(feature_extractor)
                    else :
                        assert False,("Duplicate extractors used")`

abhaykatheria commented:
-i am getting zero_divison error

marco-c commented on bugbug/bug_features.py:
```diff
@@ -557,8 +557,24 @@ def __init__(
+                used_features.append(type(feature_extractor))
+                self.feature_extractors.append(feature_extractor)
+            else:
+                assert False, "Duplicate features are used"
```
You should be able to greatly simplify the implementation by using a set:
```Python
assert len(set(type(fe) for fe in feature_extractors)) == len(feature_extractors), "Duplicate feature extractors"
self.feature_extractors = feature_extractors
```

marco-c commented:
Please also add tests to make sure the assertion is hit when it should be hit.

marco-c commented on tests/test_bug_features.py:
```diff
@@ -163,6 +165,19 @@ def test_comment_length(read):
+        url(),
+        url(),
+    ]
+    try:
```
Your code is not making sure that the assertion is thrown. You'd have to do something like:
```
try:
    BugExtractor(feature_extractors, cleanup_functions)
    assert False, "An assertion should have failed"
except AssertionError:
    pass
```

Even better, use the pytest utility to do this, like in:
https://github.com/mozilla/bugbug/blob/abf47d3b1c69e782450a57918d5950b2bec22cb4/tests/test_db.py#L82-L83

marco-c commented on tests/test_bug_features.py:
```diff
@@ -163,6 +165,19 @@ def test_comment_length(read):
     read("comment_length.json", comment_length, [566, 5291])
 
 
+def test_bug_extractor():
```
Nit: add your tests at the end of the file

marco-c replied:
A similar test should be written for CommitExtractor

marco-c commented on tests/test_bug_features.py:
```diff
@@ -163,6 +165,19 @@ def test_comment_length(read):
+        BugExtractor(feature_extractors, cleanup_functions)
+    except AssertionError:
+        print("exception handled")
+
```
There should be three tests:
1) Two feature extractors and two cleanup functions, without duplicates;
2) Two duplicate feature extractors and two non-duplicate cleanup functions;
3) Two non-duplicate feature extractors and two duplicate cleanup functions.

abhaykatheria commented:
i ran the flake 8 testing there are no changes to commit further but still this fails

abhaykatheria commented:
Sorry for the mess actually i wrote this in bit hurry and thanks for your patience.

abhaykatheria commented:
![image](https://user-images.githubusercontent.com/40055274/72011263-74cc4180-327f-11ea-898d-bb228dfd6180.png)

marco-c commented on tests/test_bug_features.py:
```diff
@@ -266,3 +268,16 @@ def test_is_same_target_milestone(test_data, expected):
+
+
+BUG_EXTRACTOR_PARAMS = [
+    ([has_str, has_url], [fileref, url]),
```
This case should not fail the assertion.

marco-c commented on tests/test_commit_features.py:
```diff
@@ -0,0 +1,24 @@
+from bugbug.feature_cleanup import fileref, url
+
+COMMIT_EXTRACTOR_PARAMS = [
+    ([has_str, has_url], [fileref, url]),
```
You should not test bug_features functions in the commit_features test

abhaykatheria commented:
![image](https://user-images.githubusercontent.com/40055274/72071455-d937e100-3311-11ea-879d-417e81babe27.png)
features i used are different but it still raises the assertion

marco-c commented:
Add some logging to figure out why:
```
print(set(type(fe) for fe in feature_extractors))
```

abhaykatheria commented:
its strange the features were read as class not objects as soon as i added the parenthesis the error went away but i did the same in test _bug_features and it ran flawlessly so i added the parenthesis everywhere to make it uniform.:)

marco-c commented on tests/test_bug_features.py:
```diff
@@ -266,3 +268,16 @@ def test_is_same_target_milestone(test_data, expected):
+
+@pytest.mark.parametrize("feature_extractors,cleanup_functions", BUG_EXTRACTOR_PARAMS)
+def test_BugExtractor(feature_extractors, cleanup_functions):
+    BugExtractor([has_str(), has_url()], [fileref(), url()])
```
You are running this twice because of the parametrization. Just remove the parametrization and run the checks directly in the test

marco-c commented on tests/test_commit_features.py:
```diff
@@ -0,0 +1,23 @@
+    "feature_extractors,cleanup_functions", COMMIT_EXTRACTOR_PARAMS
+)
+def test_CommitExtractor(feature_extractors, cleanup_functions):
+    CommitExtractor([reviewers_num(), author_experience()], [fileref(), url()])
```
Same here

marco-c commented:
> its strange the features were read as class not objects as soon as i added the parenthesis the error went away but i did the same in test _bug_features and it ran flawlessly so i added the parenthesis everywhere to make it uniform.:)

It is because they are classes, so you need to "call" them to create objects. I hadn't noticed you weren't in the test.

marco-c commented on tests/test_bug_features.py:
```diff
@@ -266,3 +268,10 @@ def test_is_same_target_milestone(test_data, expected):
+    BugExtractor([has_str(), has_url()], [fileref(), url()])
+    with pytest.raises(AssertionError):
+        BugExtractor([has_str(), has_str()], [fileref(), url()])
+        BugExtractor([has_str(), has_url()], [fileref(), fileref()])
```
This way you are only testing the first call, as it raises the assertion and exits.

You should instead do:
```
with pytest.raises(AssertionError):
    BugExtractor([has_str(), has_str()], [fileref(), url()])


with pytest.raises(AssertionError):
    BugExtractor(...)
```

marco-c commented on tests/test_commit_features.py:
```diff
@@ -0,0 +1,16 @@
+    CommitExtractor([reviewers_num(), author_experience()], [fileref(), url()])
+    with pytest.raises(AssertionError):
+        CommitExtractor([reviewers_num(), author_experience()], [fileref(), fileref()])
+        CommitExtractor([author_experience(), author_experience()], [fileref(), url()])
```
Same here

marco-c commented on bugbug/commit_features.py:
```diff
@@ -495,7 +495,17 @@ def merge_commits(commits):
     def __init__(self, feature_extractors, cleanup_functions):
+        print(feature_extractors)
+        print(type(feature_extractors[0]))
+        print(set(type(fe) for fe in feature_extractors))
```
Don't forget to remove the logging statements
